cluster/clustered_data.go

```diff
@@ -0,0 +1,199 @@
+package cluster
+
+import (
+	"encoding/binary"
+	"fmt"
+	"github.com/pkg/errors"
+	"github.com/spirit-labs/tektite/common"
+	log "github.com/spirit-labs/tektite/logger"
+	"github.com/spirit-labs/tektite/objstore"
+	"sync"
+	"sync/atomic"
+	"time"
+)
+
+/*
+ClusteredData - provides a mechanism whereby potentially large state can be loaded by a member of the cluster, updated
+and then new state stored without fear of another member overwriting the state.
+
+This is used in Tektite for persisting the controller metadata.
+
+A member of the cluster takes ownership of the state by calling `LoadState` which atomically increments an 'epoch' value
+in a `StateUpdator` instance and returns the current state, if any.
+
+The owner makes changes to the state then calls `StoreState` to persist it. This stores the state as on object in the
+object store external to the `StateUpdator`. Ideally we would store the state in the `StateUpdator` but the current S3
+implementation of conditional PUTs requires us to create a new object on each update. If the object is large this
+results in a lot of objects and a lot of storage under load.
+
+Before the call to `StoreState` returns a no-op update on `StatePersistor` occurs to read the latest epoch. If this does
+not match the epoch the state was stored with thew call returns an error.
+
+The key that we store the state under includes the epoch in it, e.g. `my-data-key-prefix-0000000234`. When data is
+loaded, the key with the highest epoch prefix is loaded.
+*/
+type ClusteredData struct {
+	lock           sync.Mutex
+	dataBucketName string
+	dataKeyPrefix  string
+	stateMachine   *StateUpdator
+	objStoreClient objstore.Client
+	epoch          uint64
+	opts           ClusteredDataOpts
+	readyState     clusteredDataState
+	stopping       atomic.Bool
+}
+
+type clusteredDataState int
+
+const (
+	clusteredDataStateReady = iota
+	clusteredDataStateLoaded
+	clusteredDataStateStopped
+)
+
+func NewClusteredData(stateMachineBucketName string, stateMachineKeyPrefix string, dataBucketName string,
+	dataKeyPrefix string, objStoreClient objstore.Client, opts ClusteredDataOpts) *ClusteredData {
+	return &ClusteredData{
+		objStoreClient: objStoreClient,
+		dataBucketName: dataBucketName,
+		dataKeyPrefix:  dataKeyPrefix,
+		stateMachine: NewStateUpdator(stateMachineBucketName, stateMachineKeyPrefix, objStoreClient,
+			StateUpdatorOpts{}),
+		opts:       opts,
+		readyState: clusteredDataStateReady,
+	}
+}
+
+type ClusteredDataOpts struct {
+	AvailabilityRetryInterval time.Duration
+	ObjStoreCallTimeout       time.Duration
+}
+
+func (mo *ClusteredDataOpts) setDefaults() {
+	if mo.AvailabilityRetryInterval == 0 {
+		mo.AvailabilityRetryInterval = DefaultAvailabilityRetryInterval
+	}
+}
+
+func (m *ClusteredData) LoadData() ([]byte, error) {
```

Comment:
> Hm, I got stuck writing here. How do we know that no other process updates the epoch in the same way this process does? They all deterministically seem to just increment it?

If a new leader starts and loads data, the epoch will be incremented. The previous leader if alive still (zombie) will still have the old epoch value, if the zombie then attempts to write, the epoch is check again, it will be found to be different and the write will fail. This is how we fence zombies.

Reply:
> The key of the stateMachine here is gonna be the controller metadata key? And the value the instance id?
But now we're using an incrementing epoch, so i'm not sure what we're doing

There's no instance id any more.

We simply store the epoch in the state updator, and we store the actual state with a key that includes the epoch in it, e.g. "my-state-prefix-0000000123"